File: src/main/GO/leetcode/lists/offer/Offer51.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/10 16:16
 */

func reversePairs(nums []int) int {
	var res = 0
	tmp := make([]int, len(nums))
	var mergeSort func(arr []int, l int, r int)
	mergeSort = func(arr []int, l int, r int) {
		if l < r {
			// Find the middle point
			m := l + (r-l)/2
			// Sort first and second halves
			mergeSort(arr, l, m)
			mergeSort(arr, m+1, r)
			// Merge the sorted halves
			i := l
			j := m + 1
			for k := l; k <= r; k++ {
				tmp[k] = nums[k]
			}
			for k := l; k <= r; k++ {
				if i == m+1 {
					nums[k] = tmp[j]
					j++
				} else if j == r+1 || tmp[i] <= tmp[j] {
					nums[k] = tmp[i]
					i++
				} else {
					nums[k] = tmp[j]
					j++
					res += m - i + 1 // 统计逆序对
				}
			}
		}
	}
	mergeSort(nums, 0, len(nums)-1)
	return res
}

File: src/main/java/leetcode/lists/offer/Offer51.java
package leetcode.lists.offer;

/**
 * @author jingxinwu
 * @date 2021-12-04 5:29 下午
 */
public class Offer51 {

    int count;

    int[] arr;

    public int reversePairs(int[] nums) {
        arr = new int[nums.length];
        mergeSort(nums, 0, nums.length - 1);
        return count;
    }

    private void mergeSort(int[] nums, int left, int right) {
        if (left >= right) {
            return;
        }
        int mid = (right - left) / 2 + left;
        mergeSort(nums, left, mid);
        mergeSort(nums, mid + 1, right);
        merge(nums, left, mid, right);
    }

    private void merge(int[] nums, int left, int mid, int right) {
        int i = left, j = mid + 1;
        int k = left;
        while (i <= mid && j <= right) {
            if (nums[i] <= nums[j]) {
                arr[k] = nums[i];
                i++;
            } else {
                arr[k] = nums[j];
                j++;
                count += mid - i + 1; // 统计逆序对
            }
            k++;
        }
        while (i <= mid) {
            arr[k] = nums[i];
            i++;
            k++;
        }
        while (j <= right) {
            arr[k] = nums[j];
            j++;
            k++;
        }
        if (right + 1 - left >= 0) System.arraycopy(arr, left, nums, left, right + 1 - left);
    }
}
